jetty-websocket/websocket-server/src/main/java/org/eclipse/jetty/websocket/server/WebSocketServerFactory.java

```diff
@@ -303,27 +313,76 @@ public Object createWebSocket(ServletUpgradeRequest req, ServletUpgradeResponse
     @Override
     protected void doStart() throws Exception
     {
-        if(this.objectFactory == null && context != null)
+        if(this.objectFactory == null)
         {
-            this.objectFactory = (DecoratedObjectFactory) context.getAttribute(DecoratedObjectFactory.ATTR);
-            if (this.objectFactory == null)
-            {
-                throw new IllegalStateException("Unable to find required ServletContext attribute: " + DecoratedObjectFactory.ATTR);
-            }
+            this.objectFactory = findDecorator();
         }
-    
-        if(this.executor == null && context != null)
+
+        if(this.executor == null)
         {
-            ContextHandler contextHandler = ContextHandler.getContextHandler(context);
-            this.executor = contextHandler.getServer().getThreadPool();
+            this.executor = findExecutor();
         }
-        
+    
         Objects.requireNonNull(this.objectFactory, DecoratedObjectFactory.class.getName());
         Objects.requireNonNull(this.executor, Executor.class.getName());
         
         super.doStart();
     }
-    
+
+    private DecoratedObjectFactory findDecorator()
+    {
+        DecoratedObjectFactory objectFactory;
+
+        if (context != null)
+        {
+            objectFactory = (DecoratedObjectFactory) context.getAttribute(DecoratedObjectFactory.ATTR);
+            if (objectFactory != null)
+            {
+                return objectFactory;
+            }
+        }
+
+        objectFactory = new DecoratedObjectFactory();
+        objectFactory.addDecorator(new DeprecationWarning());
+        LOG.info("No DecoratedObjectFactory provided, using new {}", objectFactory);
+        return objectFactory;
+    }
+
+    /**
+     * Attempt to find the Executor that should be used.
+     * @return the Executor that should be used.
+     */
+    private Executor findExecutor()
+    {
+        if(context != null)
+        {
+            // Attempt to pull Executor from ServletContext attribute
+            Executor contextExecutor = (Executor) context.getAttribute("org.eclipse.jetty.server.Executor");
```

Comment:
Is this a new attribute?  Should be a `public final static String` and documented.
Is it really the `server.Executor`?  it's more of a  `org.eclipse.jetty.context.Executor`

Reply:
This one already exists (just using it).
https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L766

Documented in javadoc here.
https://github.com/eclipse/jetty.project/blob/jetty-9.4.9.v20180320/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L104

Reply:
Heh, a blame shows that name is something you assigned back in 2013.

I would argue that an Executor is a server thing, created/owned/managed by the server.
It's not a context feature or component, that's for sure.

Even if all we have is a context, and we want the ThreadPool the most common code is ...

``` java
Executor threadPool = context.getServer().getThreadPool();
```

Reply:
Ah so it is a server threadpool, as it is the servers threadpool being exposed to the application (we should probably make that optional:)
So name is good, should have been in a public final static. can you fix?